src/sql/src/query_model/lowering.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::query_model::{
+    BoxId, BoxScalarExpr, BoxType, ColumnReference, Model, QuantifierSet, QuantifierType,
+};
+use itertools::Itertools;
+use repr::RelationType;
+use std::collections::HashMap;
+
+/// Maps a column reference to a specific column position.
+///
+/// This is used for resolving column references when lowering expressions
+/// on top of a `MirRelationExpr`, where it maps each column reference with
+/// with a column position within the projection of that `MirRelationExpr`.
+type ColumnMap = HashMap<ColumnReference, usize>;
+
+impl Model {
+    pub fn lower(&self) -> expr::MirRelationExpr {
+        let mut lowerer = Lowerer::new(self);
+        let mut id_gen = ore::id_gen::IdGen::default();
+        expr::MirRelationExpr::constant(vec![vec![]], RelationType::new(vec![]))
+            .let_in(&mut id_gen, |_id_gen, get_outer| {
+                lowerer.apply(self.top_box, get_outer, &ColumnMap::new())
+            })
+    }
+}
+
+struct Lowerer<'a> {
+    model: &'a Model,
+}
+
+impl<'a> Lowerer<'a> {
+    fn new(model: &'a Model) -> Self {
+        Self { model }
+    }
+
+    /// Applies the given box identified by its ID to the given outer relation.
+    fn apply(
+        &mut self,
+        box_id: BoxId,
+        get_outer: expr::MirRelationExpr,
```

Comment:
Is there a particular reason why the outer relation is called `get_outer` and not just `outer`?

Reply:
Because it must be `Get` around the relation of the outer context. This is to avoid cloning entire subgraphs. The same approach is used in the current lowering code from Hir to Mir. Eventually we will add the same assertion that one has.